applications/wizard.py

```diff
@@ -36,6 +36,55 @@ def get_next_page_key(self, key):
             if page.status == UNSTARTED:
                 return page.key
 
+    def is_field_optional(self, field_name, form_spec):
+        """Look up a fields optional key in the given form_spec"""
+        for fieldset in form_spec.fieldsets:
+            for field in fieldset.fields:
+                if field.name == field_name:
+                    return field.optional
+
+        # assume an unknown field is required
+        return False
+
+    def is_field_valid(self, field, form_spec):
+        """
+        Work out if a given field is valid for the benefit of the confirmation page
+
+        All fields are required to be have a value for the final confirmation step.
+        """
+        if self.is_field_optional(field.name, form_spec):
+            return True
+
+        if field.value() is None:
+            return False
+
+        if field.value() == "":
+            return False
+
+        return True
+
+    def pages_for_confirmation(self):
+        """
+        Generate a list of pages for use on the confirmation page
+
+        Build a structure for display on the confirmation page
+        """
+        for page in self.get_pages():
+            fields = [
+                {
+                    "label": field.label,
+                    "value": field.value(),
+                    "is_valid": self.is_field_valid(field, page.form_spec),
+                }
+                for field in page.get_unbound_data_form()
```

Comment:
I don't think we should be using `page.get_unbound_data_form()` here.  It's what I originally used for the confirmation page, but we've shown in the staff review page that we can render all the data for an application without using forms.

A suggestion:

1. add an `"is_valid"` key to `Field.review_context()`
2. re-write part of `confirmation.html` to use the same strategy for rendering pages as we do in [`application_detail.html`](https://github.com/opensafely-core/job-server/blob/main/staff/templates/staff/application_detail.html#L66-L82)

If you do this, I think you'll also be able to drop the `Wizard.is_field_optional` and `Wizard.is_field_valid` methods.